plugins/elantp/fu-elantp-hid-device.c

```diff
@@ -0,0 +1,474 @@
+				   NULL, error))
+		return FALSE;
+	if (rxsz == 0)
+		return TRUE;
+
+	/* GetFeature */
+	buf = g_malloc0 (rxsz + 1);
+	buf[0] = tx[0]; /* report number */
+	if (!fu_udev_device_ioctl (FU_UDEV_DEVICE (self),
+				   HIDIOCGFEATURE(rxsz + 3), buf,
+				   NULL, error))
+		return FALSE;
+
+	/* success */
+	memcpy (rx, buf + 0x3, rxsz);
+	return TRUE;
+}
+
+static gboolean
+fu_elantp_hid_device_read_cmd (FuElantpHidDevice *self, guint16 reg,
+			       guint8 *rx, gsize rxsz, GError **error)
+{
+	guint8 buf[5];
+
+	buf[0] = 0x0d; /* report number */
+	buf[1] = 0x05;
+	buf[2] = 0x03;
+	fu_common_write_uint16 (buf + 0x3, reg, G_LITTLE_ENDIAN);
+	if (g_getenv ("FWUPD_ELANTP_VERBOSE") != NULL)
+		fu_common_dump_raw (G_LOG_DOMAIN, "ReadCmd", buf, sizeof(buf));
+	return fu_elantp_hid_device_send_cmd (self, buf, sizeof(buf), rx, rxsz, error);
+}
+
+static gint
+fu_elantp_hid_device_write_cmd (FuElantpHidDevice *self,
+				guint16 reg, guint16 cmd,
+				GError **error)
+{
+	guint8 buf[5];
+
+	buf[0] = 0x0d; /* report number */
+	fu_common_write_uint16 (buf + 0x1, reg, G_LITTLE_ENDIAN);
+	fu_common_write_uint16 (buf + 0x3, cmd, G_LITTLE_ENDIAN);
+	if (g_getenv ("FWUPD_ELANTP_VERBOSE") != NULL)
+		fu_common_dump_raw (G_LOG_DOMAIN, "WriteCmd", buf, sizeof(buf));
+
+	return fu_elantp_hid_device_send_cmd (self, buf, sizeof(buf), NULL, 0, error);
+}
+
+static gboolean
+fu_elantp_hid_device_ensure_iap_ctrl (FuElantpHidDevice *self, GError **error)
+{
+	guint8 buf[2] = { 0x0 };
+	if (!fu_elantp_hid_device_read_cmd (self, ETP_CMD_I2C_IAP_CTRL, buf, sizeof(buf), error)) {
+		g_prefix_error (error, "failed to read IAPControl: ");
+		return FALSE;
+	}
+	self->iap_ctrl = fu_common_read_uint16 (buf, G_LITTLE_ENDIAN);
+
+	/* in bootloader mode? */
+	if ((self->iap_ctrl & ETP_I2C_MAIN_MODE_ON) == 0)
+		fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+	else
+		fu_device_remove_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+
+	return TRUE;
+}
+
+static gboolean
+fu_elantp_hid_device_setup (FuDevice *device, GError **error)
+{
+	FuElantpHidDevice *self = FU_ELANTP_HID_DEVICE (device);
+	FuUdevDevice *udev_device = FU_UDEV_DEVICE (device);
+	gboolean is_new_pattern;
+	guint16 fwver;
+	guint16 tmp;
+	guint8 buf[2] = { 0x0 };
+	guint8 ic_type;
+	g_autofree gchar *instance_id1 = NULL;
+	g_autofree gchar *instance_id_ic_type = NULL;
+	g_autofree gchar *version_bl = NULL;
+	g_autofree gchar *version = NULL;
+
+	/* get current firmware version */
+	if (!fu_elantp_hid_device_read_cmd (self,
+					    ETP_CMD_I2C_FW_VERSION,
+					    buf, sizeof(buf), error)) {
+		g_prefix_error (error, "failed to read fw version: ");
+		return FALSE;
+	}
+	fwver = fu_common_read_uint16 (buf, G_LITTLE_ENDIAN);
+	version = fu_common_version_from_uint16 (fwver, FWUPD_VERSION_FORMAT_HEX);
+	fu_device_set_version (device, version);
+
+	/* get EAP firmware version */
+	is_new_pattern = fu_device_has_custom_flag (FU_DEVICE (self), "new-pattern");
```

Comment:
the word new doesn't age well.  is there a better name than new?

Reply:
I used `iap-version-2`